Updating the error docs (#3512)
meshery/meshery: docs/pages/project/contributing/contributing-error.md

---

Meshery pervasively uses MeshKit as a golang and service mesh management-specific library in all of its components. MeshKit helps populate error messages with a uniform and useful set of informative attributes. 

> To help with creating error codes, MeshKit contains a tool that analyzes, verifies and updates error codes in Meshery source code trees. It extracts error details into a file that can be used for publishing all error code references on the at Meshery [error codes reference page](https://docs.meshery.io/reference/error-codes). The objective to create this was to avoid centralized handling of error codes and automating everything

Reviewer: `on the at Meshery` the `at` can be removed form here.

In order to create a Meshery error object, you will need to create a custom wrapper object for the native golang error. This can be done from the <a href="https://github.com/layer5io/meshkit/tree/master/errors">MeshKit Error</a> package. 

## Some rules for making Errors codes

- Errors names and codes are namespaced to components, i.e. they need to be unique within a component, which is verified by this tool.

- Errors are not to be reused across components and modules.

- Codes carry no meaning, as e.g. HTTP status codes do.

> - In the code, create string var's or const's with names starting with Err[A-Z] and ending in Code, e.g. 'ErrApplyManifestCode'.

Reviewer: I think this point can be updated like `For the Code argument in the errors.New use the same Error name and append a "Code" after it. e.g error name : ErrApplyManifest then the error code is ErrApplyManifestCode`.

- Set the value to any string, like "replace_me" (no convention here), e.g. ErrApplyManifestCode = "replace_me".

- Error codes are not to be set as integer 

- CI will take care of updating Error codes from a string to an integer.

- Using errors.NewDefault(...) is deprecated. This tool emits a warning if this is detected.

- Use errors.New(...) from MeshKit to create actual errors with all the details.
  This is often done in a factory function. It is important that the error code variable is used here, not a literal.

>   Specify detailed descriptions, probable causes, and remedies. They need to be string literals, call expressions are ignored.

Reviewer: And I think it can be added that the `first letter of the every error description need to capitalized`

---

Meshery pervasively uses MeshKit as a golang and service mesh management-specific library in all of its components. MeshKit helps populate error messages with a uniform and useful set of informative attributes. 

> This tool analyzes, verifies and updates error codes in Meshery source code trees. It extracts error details into a file that can be used for publishing all error code references on the Meshery website. The objective to create this was to avoid centralized handling of error codes, automating everything

Reviewer 2: ```suggestion
To help with creating error codes, MeshKit contains a tool that analyzes, verifies and updates error codes in Meshery source code trees. It extracts error details into a file that can be used for publishing all error code references at Meshery [error codes reference page](https://docs.meshery.io/reference/error-codes). The objective to create this was to avoid centralized handling of error codes, automating everything.
```

Inorder to create a meshery error object, you will need to create a custom wrapper object for the native golang error. This can be done from the <a href="https://github.com/layer5io/meshkit/tree/master/errors">Meshkit Error</a> package. 

Use the `errors.New()` function to create a new instance of the error object and pass situation-specific attributes as function arguments. 
The below are the listed attributes:
- Code
- Short Description
- Long Description
- Probable Cause
- Suggested Remediation

### Syntax
     errors.New(ErrExampleCode, errors.Alert, []string{"<short-description>"}, []string{"<long-description>"}, []string{"<probable-cause>"}, []string{"<suggested remediation>"})

> It is intended to be run locally and as part of a CI workflow.

Reviewer 2: ```suggestion

```
Maybe we don't need this here :thinking:

> 

Reviewer 2: ```suggestion
Some things to keep in mind when creating errors:
```

> - Errors names and codes are namespaced to components, i.e. they need to be unique within a component, which is verified by this tool.

Reviewer 2: These are some nice bullet points :white_check_mark:. But they seem to highlight more about the utility. I think so much info might confuse contributors.
Can these be made more like: "some rules for making error codes" and then mentioning some conventions to follow that abide by the utility?

Reviewer: `Errors names and codes` to `Errors name and code` might be?

- A component corresponds usually to a repository. Components have a type and a name. 

>   They are also returned from the ComponentInfo endpoint, e.g. for adapters.

Reviewer 2: Is there a ComponentInfo endpoint? :thinking:

Reviewer: No I think.

Reviewer 3: @DelusionalOptimist @sayantan1413 yes there is for adapters (https://github.com/meshery/meshery/blob/b04208b6614250ac723317e3040b079c93139800/meshes/meshops.proto#L17) and eventually, there should be for all components. It has a component name and type field and which should be the same  as component name and info in the component_info.json file. My idea was that this was picked up for all functionality from `component_info.json` file, so that it was guaranteed to be consistent. Eventually, component name and type should be part of any message / notification emitted by any component.

Author: @mgfeller is there anything you want me to add or remove from [here](https://60f2beb4f706dbe8151b6a6c--meshery-docs-preview.netlify.app/project/contributing-error) : )

Reviewer 3: I think this sentence could be removed: "They are also returned from the ComponentInfo endpoint, e.g. for adapters."

Reviewer 3: Otherwise, not quite sure anything should be removed or added. Personally, I always like to have enough information so that I understand how the dots connect, so I'm biased. But I understand the concern that some can be confusing. What do you think @Shreyas220 ? Anything superfluous here? 

If not, I think we can just go forward and update later...

- Use errors.New(...) from meshkit to create actual errors with all the details.
  This is often done in a factory function. It is important that the error code variable is used here, not a literal.
  Specify detailed descriptions, probable causes, and remedies. They need to be string literals, call expressions are ignored.
  This tool extracts this information from the code and exports it.
- By convention, error codes and the factory functions live in files called error.go. The tool checks all files, but updates only error.go files.
- This tool will create a couple of files, one of them is designed to be used to generate the error reference on the meshery website.
  The file errorutil_analyze_summary.json contains a summary of the analysis, notably lists of duplicates etc.
- The tool requires a file called component_info.json. Its location can be customized, by default it is the root directory (-d flag). 
  This file has the following content, with concrete values specific for each component:
  {
    "name": "meshkit",
    "type": "library",
    "next_error_code": 11010
  }
- The tool updates next_error_code.

## Example

> In this example we are trying to create an Error for not able to marhsal a JSON

Reviewer 2: ```suggestion
In this example we are trying to create an Error for being unable to marshal JSON
```

```code
var (
    // Error code 
    ErrMarshalCode= "replace_me"

    //Static errors (for example)
    ErrExample = errors.New(ErrExampleCode, errors.Alert, []string{"<short-description>"}, []string{"<long-description>"}, []string{"<probable-cause>"}, []string{"<suggested remediation>"})
)

// Dynamic errors
//Error Name
func ErrMarshal(err error, obj string) error {
	return errors.New(ErrMarshalCode, errors.Alert, []string{"Unable to marshal the : ", obj}, []string{err.Error()}, []string{}, []string{})
}

```

### Replacing old Error Codes 

 Old
 ```Code 
    bd, err := json.Marshal(providers)
	if err != nil {
		http.Error(w, "unable to marshal the providers", http.StatusInternalServerError)
		return
	}
 ```
New
  ```Code 
    bd, err := json.Marshal(providers)
    if err != nil {
            obj := "provider"
            http.Error(w, ErrMarshal(err, obj).Error(), http.StatusInternalServerError)
            return
        }
 ```

 

> ## Replacing logrus 

Reviewer 2: Some wrong indentation is prolly creating weird code blocks. https://60efe6596f7390cc570636d8--meshery-docs-preview.netlify.app/project/contributing-error#replacing-logrus

### Some rules for making Errors codes

- Errors names and codes are namespaced to components, i.e. they need to be unique within a component, which is verified by this tool.

- A component corresponds usually to a repository. Components have a type and a name. 
  They are also returned from the ComponentInfo endpoint, e.g. for adapters.
  Examples of a component types are 'adapter' and 'library', corresponding examples of names are 'istio' and 'meshkit'.

- There are no predefined error code ranges for components.
  Every component is free to use its own range, but it looks like the convention is to start at 1000.

- Errors are not to be reused across components and modules.

- Codes carry no meaning, as e.g. HTTP status codes do.

- In the code, create string var's or const's with names starting with Err[A-Z] and ending in Code, e.g. 'ErrApplyManifestCode'.

- Set the value to any string, like "replace_me" (no convention here), e.g. ErrApplyManifestCode = "test_code".

- If the value is a string, this tool will replace it with the next integer.

> - If the value is an int, e.g. ErrGetName = "1000" the tool will not replace it unless it is forced (command line flag --force).

Reviewer 3: It should be `ErrGetNameCode = "1000"` (my bad).

- Use errors.New(...) from meshkit to create actual errors with all the details.
  This is often done in a factory function. It is important that the error code variable is used here, not a literal.
  Specify detailed descriptions, probable causes, and remedies. They need to be string literals, call expressions are ignored.
  This tool extracts this information from the code and exports it.

- By convention, error codes and the factory functions live in files called error.go. The tool checks all files, but updates only error.go files.

- This tool will create a couple of files, one of them is designed to be used to generate the error reference on the meshery website.
  The file errorutil_analyze_summary.json contains a summary of the analysis, notably lists of duplicates etc.

- The tool requires a file called component_info.json. Its location can be customized, by default it is the root directory (-d flag). 
  This file has the following content, with concrete values specific for each component:
  {
    "name": "meshkit",
    "type": "library",
    "next_error_code": 11010
  }

- The tool updates next_error_code.

## Example

> In this example we are trying to create an Error for being unable to marshal JSON

Reviewer 3: 'we are creating an Error' not just trying :smile:

)

// Dynamic errors
//Error Name
func ErrMarshal(err error, obj string) error {
	return errors.New(ErrMarshalCode, errors.Alert, []string{"Unable to marshal the : ", obj}, []string{err.Error()}, []string{}, []string{})
}

```

### Replacing old Error Codes 

 Old
 ```Code 
    bd, err := json.Marshal(providers)
	if err != nil {
		http.Error(w, "unable to marshal the providers", http.StatusInternalServerError)
		return
	}
 ```
New
  ```Code 
    bd, err := json.Marshal(providers)
    if err != nil {
            obj := "provider"
            http.Error(w, ErrMarshal(err, obj).Error(), http.StatusInternalServerError)
            return
        }
 ```

 
## Replacing logrus 
 There already exists an interface for logger in <a href="https://github.com/layer5io/meshkit/blob/master/logger/logger.go">Meshkit </a>  

#### Defining a Logger 

   ```Code 
    type Logger struct {
        log   logger.Handler
    }
 ```

#### Debug
 
 Old

    `logrus.Debugf("meshLocationURL: %s", meshLocationURL)`
 New

    `l.log.Debug("meshLocationURL: ", meshLocationURL)`

#### Error

##### Old
  `logrus.Errorf("error marshaling data: %v.", err)`
##### New
  `l.log.Error(ErrMarshal(err, obj))`

> 

Reviewer: We can also add the `logrus.logf` to `l.log.log`.

> 

Reviewer: We can also add the `logrus.Infof` to `l.log.Info`

Author: i havnt used it yet, could write the syntax for it
